File: 임현규/programmers/Lv2/lv2_메뉴리뉴얼_하.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;
import java.util.function.Function;
import java.util.stream.Collectors;

class Solution {

    public String[] solution(String[] orders, int[] course) {
        List<String> answer = new ArrayList<>();
        for (int r : course) {
            List<String> temp = new ArrayList<>();
            for (String order : orders) {
                String ordered = Arrays.stream(order.split("")).sorted()
                    .collect(Collectors.joining());
                if (ordered.length() >= r) {
                    Combinations.combinations(ordered, new char[r], 0, 0, r, temp);
                }
            }
            if (temp.isEmpty()) continue;
            
            Map<String, Long> frequency = temp.stream()
                .collect(Collectors.groupingBy(Function.identity(), Collectors.counting()));

            int maxValue = frequency.values().stream().mapToInt(Long::intValue).max()
                .orElseThrow();

            frequency.entrySet().stream()
                .filter(entry -> entry.getValue() > 1 && entry.getValue() == maxValue)
                .map(Entry::getKey).forEach(key -> answer.add(key));
        }
        return answer.stream().sorted().toArray(String[]::new);
    }
}


class Combinations {

    public static void combinations(String s, char[] picked, int pos, int step, int r,
        List<String> answer) {
        if (step == r) {
            answer.add(String.valueOf(picked));
            return;
        }

        for (int i = pos; i < s.length(); ++i) {
            picked[step] = s.charAt(i);
            combinations(s, picked, i + 1, step + 1, r, answer);
        }
    }

}

File: 임현규/programmers/Lv2/lv2_메뉴_리뉴얼_하.py
""" 
1. 조합을 이용해 course 수만큼 담는다.
2. 구한 조합의 갯수를 센다.
3. 최대 갯수를 구한 후 중복되는 값이 있다면 모두 넣는다.(단, 갯수가 2번 이상이여야 한다.)
4. 결과를 사전순으로 정렬해서 출력한다.

"""

from collections import Counter
from itertools import combinations



def solution(orders, course):
    result = []
    for r in course:
        possibility = []
        for order in orders:
            comb = combinations(sorted(order), r)
            possibility += comb
        if not possibility: continue
        counter = Counter(possibility)
        max_value = counter.most_common(1)[0][1]
        result.extend(
            ["".join(key) for key, value in counter.items() if value == max_value and value > 1] 
        )
        result.sort()
    return result


print(solution(["XYZ", "XWY", "WXA"], [2,3,4]))

